File: web/src/services/notificationService.js
import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080/api';

const getAuthHeaders = () => {
  const headers = { 'Content-Type': 'application/json' };
  const token = localStorage.getItem('token');
  if (token) headers['Authorization'] = `Bearer ${token}`;
  return headers;
};

export const notificationService = {
  fetchUserNotifications: async (schoolId, userRole) => {
    try {
      const response = await axios.get(
        `${API_URL}/notifications/user/${encodeURIComponent(schoolId)}/role/${encodeURIComponent(userRole)}`,
        { headers: getAuthHeaders() }
      );
      return { success: true, data: Array.isArray(response.data) ? response.data : [] };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to fetch notifications',
        data: []
      };
    }
  },

  markAsRead: async (notificationId) => {
    try {
      const response = await axios.put(
        `${API_URL}/notifications/${notificationId}/read`,
        {},
        { headers: getAuthHeaders() }
      );
      return { success: true, data: response.data };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to mark as read'
      };
    }
  },

  deleteNotification: async (notificationId) => {
    try {
      await axios.delete(
        `${API_URL}/notifications/${notificationId}`,
        { headers: getAuthHeaders() }
      );
      return { success: true };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to delete notification'
      };
    }
  },

  sendNotificationToAllStudents: async (title, message) => {
    try {
      const response = await axios.post(
        `${API_URL}/notifications/broadcast/students`,
        { title, message },
        { headers: getAuthHeaders() }
      );
      return { success: true, data: response.data };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to send notification to students'
      };
    }
  },

  sendNotificationToEveryone: async (title, message) => {
    try {
      const response = await axios.post(
        `${API_URL}/notifications/broadcast/all`,
        { title, message },
        { headers: getAuthHeaders() }
      );
      return { success: true, data: response.data };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to send notification to everyone'
      };
    }
  },

  sendNotificationToUser: async (schoolId, title, message, type = 'info') => {
    try {
      const response = await axios.post(
        `${API_URL}/notifications/send`,
        { recipientId: schoolId, title, message, type },
        { headers: getAuthHeaders() }
      );
      return { success: true, data: response.data };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to send notification'
      };
    }
  },

  sendNotificationToAllStaff: async (title, message) => {
    try {
      const response = await axios.post(
        `${API_URL}/notifications/broadcast/staff`,
        { title, message },
        { headers: getAuthHeaders() }
      );
      return { success: true, data: response.data };
    } catch (error) {
      return {
        success: false,
        error: error.response?.data?.error || error.message || 'Failed to send notification to staff'
      };
    }
  }
};

export default notificationService;

File: mobile/app/src/main/java/edu/cit/bayonas/citmedconnect/data/api/NotificationService.kt
package edu.cit.bayonas.citmedconnect.data.api

import edu.cit.bayonas.citmedconnect.data.model.BroadcastNotificationRequest
import edu.cit.bayonas.citmedconnect.data.model.NotificationResponse
import edu.cit.bayonas.citmedconnect.data.model.UnreadCountResponse
import retrofit2.http.*

interface NotificationService {

    @GET("api/notifications/user/{schoolId}/role/{role}")
    suspend fun getNotificationsForUser(
        @Path("schoolId") schoolId: String,
        @Path("role") role: String
    ): List<NotificationResponse>

    @GET("api/notifications/user/{schoolId}/unread-count")
    suspend fun getUnreadCount(
        @Path("schoolId") schoolId: String
    ): UnreadCountResponse

    @PUT("api/notifications/{id}/read")
    suspend fun markAsRead(
        @Path("id") id: String
    ): NotificationResponse

    @DELETE("api/notifications/{id}")
    suspend fun deleteNotification(
        @Path("id") id: String
    )

    @POST("api/notifications/broadcast/students")
    suspend fun broadcastToStudents(
        @Body request: BroadcastNotificationRequest
    ): List<NotificationResponse>

    @POST("api/notifications/broadcast/staff")
    suspend fun broadcastToStaff(
        @Body request: BroadcastNotificationRequest
    ): List<NotificationResponse>

    @POST("api/notifications/broadcast/all")
    suspend fun broadcastToAll(
        @Body request: BroadcastNotificationRequest
    ): List<NotificationResponse>
}
